### ISSUE TITLE
Can't transpile more than one circuit at the same time on MacOs Apple Silicon

### ISSUE THREAD

**Issue by muuusso** [Qiskit/qiskit#11419]
### Environment

- **Qiskit Terra version**:  0.45.0
- **Python version**:  3.10
- **Operating system**: MacOs (Apple Silicon)

### What is happening?

If I try to pass a list of circuits to the transpile function I get a pickle error. 

### How can we reproduce the issue?

```python
from qiskit import QuantumCircuit, transpile
from qiskit_aer import AerSimulator

N = 4
state1 = QuantumCircuit(N)
state2 = QuantumCircuit(N)

simulator = AerSimulator()

transpile([state1, state2], simulator)
```
<details>
<summary> If I run the simple code above I obtain the rather lenghty following error</summary>

```python
TypeError                                 Traceback (most recent call last)
Cell In[3], line 10
      6 state2 = QuantumCircuit(N)
      8 simulator = AerSimulator()
---> 10 transpile([state1, state2], simulator)

File ~/thesis/code/.venv/lib/python3.10/site-packages/qiskit/compiler/transpiler.py:428, in transpile(circuits, backend, basis_gates, inst_map, coupling_map, backend_properties, initial_layout, layout_method, routing_method, translation_method, scheduling_method, instruction_durations, dt, approximation_degree, timing_constraints, seed_transpiler, optimization_level, callback, output_name, unitary_synthesis_method, unitary_synthesis_plugin_config, target, hls_config, init_method, optimization_method, ignore_backend_supplied_default_methods)
    400         instruction_durations = _parse_instruction_durations(
    401             backend, instruction_durations, dt, circuits[0]
    402         )
    404 pm = generate_preset_pass_manager(
    405     optimization_level,
    406     backend=backend,
   (...)
    426     _skip_target=_skip_target,
    427 )
--> 428 out_circuits = pm.run(circuits, callback=callback)
    429 for name, circ in zip(output_name, out_circuits):
    430     circ.name = name

File ~/thesis/code/.venv/lib/python3.10/site-packages/qiskit/transpiler/passmanager.py:532, in StagedPassManager.run(self, circuits, output_name, callback)
    525 def run(
    526     self,
    527     circuits: _CircuitsT,
    528     output_name: str | None = None,
    529     callback: Callable | None = None,
    530 ) -> _CircuitsT:
    531     self._update_passmanager()
--> 532     return super().run(circuits, output_name, callback)

File ~/thesis/code/.venv/lib/python3.10/site-packages/qiskit/transpiler/passmanager.py:551, in _replace_error.<locals>.wrapper(*meth_args, **meth_kwargs)
    548 @wraps(meth)
    549 def wrapper(*meth_args, **meth_kwargs):
    550     try:
--> 551         return meth(*meth_args, **meth_kwargs)
    552     except PassManagerError as ex:
    553         raise TranspilerError(ex.message) from ex

File ~/thesis/code/.venv/lib/python3.10/site-packages/qiskit/transpiler/passmanager.py:291, in PassManager.run(self, circuits, output_name, callback)
    288 if callback is not None:
    289     callback = _legacy_style_callback(callback)
--> 291 return super().run(
    292     in_programs=circuits,
    293     callback=callback,
    294     output_name=output_name,
    295 )

File ~/thesis/code/.venv/lib/python3.10/site-packages/qiskit/passmanager/passmanager.py:242, in BasePassManager.run(***failed resolving arguments***)
    233 del kwargs
    235 # Pass manager may contain callable and we need to serialize through dill rather than pickle.
    236 # See https://github.com/Qiskit/qiskit-terra/pull/3290
    237 # Note that serialized object is deserialized as a different object.
    238 # Thus, we can resue the same manager without state collision, without building it per thread.
    239 return parallel_map(
    240     _run_workflow_in_new_process,
    241     values=in_programs,
--> 242     task_kwargs={"pass_manager_bin": dill.dumps(self)},
    243 )

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:263, in dumps(obj, protocol, byref, fmode, recurse, **kwds)
    239 """
    240 Pickle an object to a string.
    241 
   (...)
    260 Default values for keyword arguments can be set in :mod:`dill.settings`.
    261 """
    262 file = StringIO()
--> 263 dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
    264 return file.getvalue()

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:235, in dump(obj, file, protocol, byref, fmode, recurse, **kwds)
    233 _kwds = kwds.copy()
    234 _kwds.update(dict(byref=byref, fmode=fmode, recurse=recurse))
--> 235 Pickler(file, protocol, **_kwds).dump(obj)
    236 return

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:394, in Pickler.dump(self, obj)
    392 def dump(self, obj): #NOTE: if settings change, need to update attributes
    393     logger.trace_setup(self)
--> 394     StockPickler.dump(self, obj)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:487, in _Pickler.dump(self, obj)
    485 if self.proto >= 4:
    486     self.framer.start_framing()
--> 487 self.save(obj)
    488 self.write(STOP)
    489 self.framer.end_framing()

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:932, in _Pickler.save_list(self, obj)
    929     self.write(MARK + LIST)
    931 self.memoize(obj)
--> 932 self._batch_appends(obj)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:956, in _Pickler._batch_appends(self, items)
    954     write(MARK)
    955     for x in tmp:
--> 956         save(x)
    957     write(APPENDS)
    958 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:932, in _Pickler.save_list(self, obj)
    929     self.write(MARK + LIST)
    931 self.memoize(obj)
--> 932 self._batch_appends(obj)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:956, in _Pickler._batch_appends(self, items)
    954     write(MARK)
    955     for x in tmp:
--> 956         save(x)
    957     write(APPENDS)
    958 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:887, in _Pickler.save_tuple(self, obj)
    885 if n <= 3 and self.proto >= 2:
    886     for element in obj:
--> 887         save(element)
    888     # Subtle.  Same as in the big comment below.
    889     if id(obj) in memo:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1698, in save_type(pickler, obj, postproc_list)
   1696             postproc_list = []
   1697         postproc_list.append((setattr, (obj, '__qualname__', obj_name)))
-> 1698     _save_with_postproc(pickler, (_create_type, (
   1699         type(obj), obj.__name__, obj.__bases__, _dict
   1700     )), obj=obj, postproc_list=postproc_list)
   1701     logger.trace(pickler, "# T2")
   1702 else:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1070, in _save_with_postproc(pickler, reduction, is_pickler_dill, obj, postproc_list)
   1067     pickler._postproc[id(obj)] = postproc_list
   1069 # TODO: Use state_setter in Python 3.8 to allow for faster cPickle implementations
-> 1070 pickler.save_reduce(*reduction, obj=obj)
   1072 if is_pickler_dill:
   1073     # pickler.x -= 1
   1074     # print(pickler.x*' ', 'pop', obj, id(obj))
   1075     postproc = pickler._postproc.pop(id(obj))

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:692, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    690 else:
    691     save(func)
--> 692     save(args)
    693     write(REDUCE)
    695 if obj is not None:
    696     # If the object is already in the memo, this means it is
    697     # recursive. In this case, throw away everything we put on the
    698     # stack, and fetch the object back from the memo.

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:902, in _Pickler.save_tuple(self, obj)
    900 write(MARK)
    901 for element in obj:
--> 902     save(element)
    904 if id(obj) in memo:
    905     # Subtle.  d was not in memo when we entered save_tuple(), so
    906     # the process of saving the tuple's elements must have saved
   (...)
    910     # could have been done in the "for element" loop instead, but
    911     # recursive tuples are a rare thing.
    912     get = self.get(memo[id(obj)][0])

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:578, in _Pickler.save(self, obj, save_persistent_id)
    576 reduce = getattr(obj, "__reduce_ex__", None)
    577 if reduce is not None:
--> 578     rv = reduce(self.proto)
    579 else:
    580     reduce = getattr(obj, "__reduce__", None)

TypeError: cannot pickle '_abc._abc_data' object
```
</details>

### What should happen?

It should run in parallel the transpilation

**Comment by jakelishman** [comment]
This is very odd, thanks for the report.  Could you check what versions of `qiskit_aer` and `dill` you have installed?  As a follow-up, can you try doing:
```python
import dill
import qiskit_aer

print(dill.__version__)
print(qiskit_aer.__version__)

simulator = qiskit_aer.AerSimulator()
dill.dumps(simulator)
```
and see if that produces any output, or a similar error to before?

**Comment by muuusso** [comment]
The `dill` version is `0.3.6` and `qiskit_aer` version is `0.13.1`.
 
<details>
<summary>The output for code is an error very similar to the one before</summary>

```python
TypeError                                 Traceback (most recent call last)
Cell In[9], line 8
      5 print(qiskit_aer.__version__)
      7 simulator = qiskit_aer.AerSimulator()
----> 8 dill.dumps(simulator)

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:263, in dumps(obj, protocol, byref, fmode, recurse, **kwds)
    239 """
    240 Pickle an object to a string.
    241 
   (...)
    260 Default values for keyword arguments can be set in :mod:`dill.settings`.
    261 """
    262 file = StringIO()
--> 263 dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
    264 return file.getvalue()

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:235, in dump(obj, file, protocol, byref, fmode, recurse, **kwds)
    233 _kwds = kwds.copy()
    234 _kwds.update(dict(byref=byref, fmode=fmode, recurse=recurse))
--> 235 Pickler(file, protocol, **_kwds).dump(obj)
    236 return

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:394, in Pickler.dump(self, obj)
    392 def dump(self, obj): #NOTE: if settings change, need to update attributes
    393     logger.trace_setup(self)
--> 394     StockPickler.dump(self, obj)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:487, in _Pickler.dump(self, obj)
    485 if self.proto >= 4:
    486     self.framer.start_framing()
--> 487 self.save(obj)
    488 self.write(STOP)
    489 self.framer.end_framing()

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1698, in save_type(pickler, obj, postproc_list)
   1696             postproc_list = []
   1697         postproc_list.append((setattr, (obj, '__qualname__', obj_name)))
-> 1698     _save_with_postproc(pickler, (_create_type, (
   1699         type(obj), obj.__name__, obj.__bases__, _dict
   1700     )), obj=obj, postproc_list=postproc_list)
   1701     logger.trace(pickler, "# T2")
   1702 else:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1070, in _save_with_postproc(pickler, reduction, is_pickler_dill, obj, postproc_list)
   1067     pickler._postproc[id(obj)] = postproc_list
   1069 # TODO: Use state_setter in Python 3.8 to allow for faster cPickle implementations
-> 1070 pickler.save_reduce(*reduction, obj=obj)
   1072 if is_pickler_dill:
   1073     # pickler.x -= 1
   1074     # print(pickler.x*' ', 'pop', obj, id(obj))
   1075     postproc = pickler._postproc.pop(id(obj))

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:692, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    690 else:
    691     save(func)
--> 692     save(args)
    693     write(REDUCE)
    695 if obj is not None:
    696     # If the object is already in the memo, this means it is
    697     # recursive. In this case, throw away everything we put on the
    698     # stack, and fetch the object back from the memo.

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:902, in _Pickler.save_tuple(self, obj)
    900 write(MARK)
    901 for element in obj:
--> 902     save(element)
    904 if id(obj) in memo:
    905     # Subtle.  d was not in memo when we entered save_tuple(), so
    906     # the process of saving the tuple's elements must have saved
   (...)
    910     # could have been done in the "for element" loop instead, but
    911     # recursive tuples are a rare thing.
    912     get = self.get(memo[id(obj)][0])

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:972, in _Pickler.save_dict(self, obj)
    969     self.write(MARK + DICT)
    971 self.memoize(obj)
--> 972 self._batch_setitems(obj.items())

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:998, in _Pickler._batch_setitems(self, items)
    996     for k, v in tmp:
    997         save(k)
--> 998         save(v)
    999     write(SETITEMS)
   1000 elif n:

File ~/thesis/code/.venv/lib/python3.10/site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File /opt/homebrew/Cellar/python@3.10/3.10.13_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py:578, in _Pickler.save(self, obj, save_persistent_id)
    576 reduce = getattr(obj, "__reduce_ex__", None)
    577 if reduce is not None:
--> 578     rv = reduce(self.proto)
    579 else:
    580     reduce = getattr(obj, "__reduce__", None)

TypeError: cannot pickle '_abc._abc_data' object
```
</details>

**Comment by jakelishman** [comment]
Ah nice, thanks - I think this might be a bug in the version of `dill` you have.  If you update `dill` to 0.3.7 at least, I think hopefully it'll work.

**Comment by muuusso** [comment]
I updated `dill` and now everything works. Maybe the requirements.txt should be updated to reflect this.

Thank you very much for your help.

**Comment by jakelishman** [comment]
Updating the requirement is a shade tricky, because the component that's actually the part failing to serialise here is `qiskit_aer` starting from version 0.13 (its latest), which doesn't use `dill` itself and isn't a dependency of Qiskit.

I'm a little hesitant to jump Qiskit's version requirement to "only the exact latest" of `dill`, when none of our package internals need it; it could easily make us incompatible with other packages if there are unrelated issues in dill 0.3.7 that have caused others to exclude that version.  If this error turns out to be causing lots of people problems we can potentially revisit that, or if `dill` increase their release cadence, but right now I think it's a little safer for us to trade off in favour of being more open with our requirements.

I'll close the issue as "solved" now, but if there's more, please feel free to open a new issue (or if anybody else sees this with a similar problem, and wants to revisit the `dill`-dependency question).
